pkg/soc/soc.go

```diff
@@ -0,0 +1,264 @@
+	address []byte
+}
+
+// NewOwner creates a new Owner.
+func NewOwner(address []byte) (*Owner, error) {
+	if len(address) != AddressSize {
+		return nil, fmt.Errorf("invalid address %x", address)
+	}
+	return &Owner{
+		address: address,
+	}, nil
+}
+
+// Chunk wraps a single soc update.
+type Chunk struct {
+	id        Id
+	span      int64
+	payload   []byte
+	signature []byte
+	signer    crypto.Signer
+	owner     *Owner
+}
+
+// NewChunk creates a new Chunk from arbitrary soc id and
+// soc byte payload.
+//
+// By default the span of the soc data is set to the length
+// of the payload.
+func NewChunk(id Id, payload []byte) *Chunk {
+	return &Chunk{
+		id:      id,
+		payload: payload,
+		span:    int64(len(payload)),
+	}
+}
+
+// WithSpan is a chainable function that adds an arbitrary
+// span to the chunk data defined for the update.
+func (s *Chunk) WithSpan(span int64) *Chunk {
+	s.span = span
+	return s
+}
+
+// WithOwnerAddress provides the possibility of setting the ethereum
+// address for the owner of an update in the absence of a signer.
+func (s *Chunk) WithOwnerAddress(ownerAddress *Owner) *Chunk {
+	s.owner = ownerAddress
+	return s
+}
+
+// AddSigner currently sets a single signer for the soc update.
+//
+// This method will overwrite any value set with WithOwnerAddress with
+// the address derived from the given signer.
+func (s *Chunk) AddSigner(signer crypto.Signer) error {
+	publicKey, err := signer.PublicKey()
+	if err != nil {
+		return err
+	}
+	ownerAddressBytes, err := crypto.NewEthereumAddress(*publicKey)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(os.Stderr, "%x", ownerAddressBytes)
+	ownerAddress, err := NewOwner(ownerAddressBytes)
+	if err != nil {
+		return err
+	}
+	s.signer = signer
+	s.owner = ownerAddress
+	return nil
+}
+
+// OwnerAddress returns the ethereum address of the signer of the Chunk.
+func (s *Chunk) OwnerAddress() []byte {
+	return s.owner.address
+}
+
+// Address returns the soc Chunk address of the update.
+func (s *Chunk) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
+}
+
+// FromChunk recreates an Chunk from swarm.Chunk data.
+func FromChunk(ch swarm.Chunk) (*Chunk, error) {
+	chunkData := ch.Data()
+	if len(chunkData) < minChunkSize {
+		return nil, errors.New("less than minimum length")
+	}
+
+	// add all the data fields
+	sch := &Chunk{}
+	cursor := 0
+
+	sch.id = chunkData[cursor : cursor+IdSize]
+	cursor += IdSize
+
+	sch.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
+
+	spanBytes := chunkData[cursor : cursor+swarm.SpanSize]
+	span := binary.LittleEndian.Uint64(spanBytes)
+	sch.span = int64(span)
+	cursor += swarm.SpanSize
+
+	sch.payload = chunkData[cursor:]
```

Comment:
if len(chunkData) < (minChunkSize + sch.span) {
    error
}

Reply:
I'm not sure I understand. Minimum length is checked in L128.

Reply:
after we get the spanSize, is it not good to check if we have data for that span?